### PR TITLE
[LOOP-2169] UX updates to the insulin delivery log

### PR DESCRIPTION
**Pull request by nhamming**
https://tidepool.atlassian.net/browse/LOOP-2169

UX updates to the insulin delivery log

### CODE REVIEW

File: LoopKitUI/InsulinKit/InsulinDeliveryTableViewController.swift [hunk 1]
```diff
@@ -16,15 +16,30 @@ public final class InsulinDeliveryTableViewController: UITableViewController {
 
     @IBOutlet var needsConfigurationMessageView: ErrorBackgroundView!
 
-    @IBOutlet weak var iobValueLabel: UILabel!
+    @IBOutlet weak var iobValueLabel: UILabel! {
+        didSet {
+            iobValueLabel.textColor = headerValueLabelColor
```
Comment (darinkrauss) [inline LoopKitUI/InsulinKit/InsulinDeliveryTableViewController.swift:21]: [nit] Why not just use `.label` directly as you do for `cell.textLabel?.textColor` on line 384 below?

File: LoopKit/InsulinKit/DoseType.swift [hunk 2]
```diff
@@ -15,6 +15,21 @@ public enum DoseType: String, CaseIterable {
     case resume
     case suspend
     case tempBasal
+    
+    var localizedDescription: String {
+        switch self {
+        case .basal:
+            return NSLocalizedString("Basal", comment: "localized term for basal dose type")
```
Comment (darinkrauss) [inline LoopKit/InsulinKit/DoseType.swift]: [nit] Most of the localized string comments start with a capitalized word and don't mention anything about localization. Perhaps just "Title for basal dose type"?

File: LoopKit/Extensions/DoseEntry+LocalizedDescription.swift [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    static var units = HKUnit.internationalUnit()
+
+    private var numberFormatter: NumberFormatter {
+        let numberFormatter = NumberFormatter()
+        numberFormatter.maximumFractionDigits = DoseEntry.unitsPerHour.maxFractionDigits
+        return numberFormatter
+    }
+
+    public var localizedAttributedDescription: NSAttributedString {
+        let typeAndUnitAttributes: [NSAttributedString.Key: Any] = [
```
Comment (darinkrauss) [inline LoopKit/Extensions/DoseEntry+LocalizedDescription.swift]: [nit] Isn't this making a lot of assumptions about where it is going to be displayed? Shouldn't this be in the relevant UI code?

Comment (darinkrauss) [reply LoopKit/Extensions/DoseEntry+LocalizedDescription.swift]: I'd be fine with this if the extension was fileprivate in the view or view model file.

File: LoopKit/Extensions/DoseEntry+LocalizedDescription.swift [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        case .basal, .bolus, .tempBasal:
+            let unitString = type == .bolus ? DoseEntry.units.shortLocalizedUnitString() : DoseEntry.unitsPerHour.shortLocalizedUnitString()
+            let value: Double = type == .bolus ? (deliveredUnits ?? programmedUnits) : unitsPerHour
+            
+            let valueAttributes: [NSAttributedString.Key: Any] = [
+                .font: UIFont.preferredFont(forTextStyle: .headline),
+                .foregroundColor: UIColor.label
+            ]
+             
+            let typeAttributedString = NSAttributedString(string: type.localizedDescription + ": ", attributes: typeAndUnitAttributes)
```
Comment (darinkrauss) [inline LoopKit/Extensions/DoseEntry+LocalizedDescription.swift]: I think all of this needs to go in a localized string to cover left-to-right ordering and such.

**Review by darinkrauss — COMMENTED** [review]
I have some concerns about the attributed localized dose entry.

File: LoopKitUI/InsulinKit/InsulinDeliveryTableViewController.swift (hunk 1 above)
Comment (nhamming) [reply LoopKitUI/InsulinKit/InsulinDeliveryTableViewController.swift:21]: This colour is now configurable by Loop

File: LoopKit/InsulinKit/DoseEntry.swift [hunk 5]
```diff
@@ -46,7 +46,9 @@ public struct DoseEntry: TimelineValue, Equatable {
 
 
 extension DoseEntry {
-    static let unitsPerHour = HKUnit.internationalUnit().unitDivided(by: .hour())
+    static public var units = HKUnit.internationalUnit()
```
Comment (darinkrauss) [inline LoopKit/InsulinKit/DoseEntry.swift]: [nit] `public static`?

File: LoopKitUI/InsulinKit/InsulinDeliveryTableViewController.swift [hunk 6]
```diff
@@ -365,14 +381,15 @@ public final class InsulinDeliveryTableViewController: UITableViewController {
...
+                cell.textLabel?.textColor = .label
                 cell.detailTextLabel?.text = time
                 cell.accessoryType = .none
                 cell.selectionStyle = .none
             case .history(let values):
                 let entry = values[indexPath.row]
                 let time = timeFormatter.string(from: entry.date)
 
-                cell.textLabel?.text = entry.title ?? LocalizedString("Unknown", comment: "The default title to use when an entry has none")
+                cell.textLabel?.attributedText = entry.dose?.localizedAttributedDescription ?? NSAttributedString(string: LocalizedString("Unknown", comment: "The default description to use when an entry has no dose description"))
```
Comment (darinkrauss) [inline LoopKitUI/InsulinKit/InsulinDeliveryTableViewController.swift]: [nit] What do you think about:
```
if let attributedText = entry.dose?.localizedAttributedDescription {
  cell.textLabel?.attributedText = attributedText
} else {
  cell.textLabel?.text = LocalizedString("Unknown", comment: "The default description to use when an entry has no dose description")
}
```
This seems more accurate and gets rid of the artificial `NSAttributedString(string:)` wrapper.

File: LoopKitUI/InsulinKit/InsulinDeliveryTableViewController.swift [hunk 7]
```diff
@@ -365,14 +381,15 @@ public final class InsulinDeliveryTableViewController: UITableViewController {
                 let time = timeFormatter.string(from: entry.startDate)
 
                 cell.textLabel?.text = "\(volume) U"
```
Comment (darinkrauss) [inline LoopKitUI/InsulinKit/InsulinDeliveryTableViewController.swift]: [nit] I know this wasn't changed in this PR, but should this be a localized string?

File: LoopKitUI/InsulinKit/InsulinDeliveryTableViewController.swift [hunk 8]
```diff
@@ -475,3 +492,48 @@ fileprivate extension UIAlertController {
...
+    fileprivate var localizedAttributedDescription: NSAttributedString? {
+        let font = UIFont.preferredFont(forTextStyle: .body)
+
+        switch type {
+        case .basal, .bolus, .tempBasal:
+            let unitString = type == .bolus ? DoseEntry.units.shortLocalizedUnitString() : DoseEntry.unitsPerHour.shortLocalizedUnitString()
+            let value: Double = type == .bolus ? (deliveredUnits ?? programmedUnits) : unitsPerHour
+
+            let description = String(format: NSLocalizedString("%1$@: <b>%2$@</b> %3$@", comment: "Description of a basal, bolus, or temp basal dose entry (1: title for dose type, 2: value (? if no value) in bold, 3: unit)"), type.localizedDescription, numberFormatter.string(from: value) ?? "?", unitString)
+            let descriptionWithFont = String(format:"<style>body{font-family: '-apple-system', '\(font.fontName)'; font-size: \(font.pointSize);}</style>%@", description)
```
Comment (darinkrauss) [inline LoopKitUI/InsulinKit/InsulinDeliveryTableViewController.swift:518]: [nit] I think the above usage of "<b>...</b>" is simple and straightforward enough, but this line seems overly complicated. Can we use the first line and then wrap with the previous attributed text code to wrap it with the font settings? After all, the font settings apply to all of the text while the bold only applies to the value.

**Review by darinkrauss — APPROVED** [review]
A couple of comments/questions, but nothing blocking approval. 🎃

File: LoopKitUI/InsulinKit/InsulinDeliveryTableViewController.swift (hunk 8 above)
Comment (nhamming) [reply LoopKitUI/InsulinKit/InsulinDeliveryTableViewController.swift:518]: I think I understand your question. Unfortunately, applying the font using the typical methods overrides the `bold` tag.

**Pull request merged**